### DISCUSSION TITLE
webui docker-compose build시 python version인식 Error

### DISCUSSION THREAD

**Discussion by TeusEE** [lablup/contribution-academy-2021#76] [Q&A] [answered]
**Describe the problem**
webui 실습을 위해 backend.ai-webui의 build를 위해서 build 진행 중 
https://github.com/lablup/backend.ai-webui#web-server
해당 위치까지 완료 후 docker-compose build과정에서 Python버전 missmatch오류 발생

![image](https://user-images.githubusercontent.com/79349894/130327478-2f791a1b-8c9a-4158-9f5b-7739d69fe6ef.png)

해당 문제를 해결하고자 Python3.X 를 Ubnutn 내에서 모두 삭제
-> 여전히 WSL2.0의 Default Python version(3.8.10)으로 인식하여 build과정에서 Error발생

해당 폴더를 pyenv 3.9.5가상화 환경으로 설정 후 build 진행
-> 여전히 WSL2.0의 Default Python version(3.8.10)으로 인식하여 build과정에서 Error발생

**To Reproduce**
WSL2.0 -> pyenv 설치 -> pyenv install 3.9.5 -> sudo apt-get purge --auto-remove python3 로 Ubuntu 내 python3 버전 제거
-> backend.ai-webui에서 3.9.5 환경 local로 설정 -> node js & npm 설치 -> npm i & make clean -> backend.ai-webui git clone -> backend.ai-webui 내에 config.toml 생성 ->빌드 진행

**Expected behavior**
정상적인 빌드 성공

**Screenshots**
![image](https://user-images.githubusercontent.com/79349894/130327762-b0499bb8-286f-4535-8cff-83b4572d9973.png)
Ubuntu 내 python환경 확인 결과 3.8은 더이상 존재하지 않음

![image](https://user-images.githubusercontent.com/79349894/130327796-a07e312a-97d0-4efd-a200-b75b0e09e7ea.png)
pyenv 내 버전 내에서도 3.8버전이 없음

Environment Description
OS : WSL2 Ubuntu 20.04 LTS
Current shell : Bash

**Comment by youngjun0627** [comment, marked as answer]
Dockerfile을 통해서 환경을 구축하는데 단순히 python3으로만 되어있어서 3.8.10으로 다운되는거 같아요. Dockerfile에서 python3 대신 python3.9로 고치고 실행해보시겠어요?
예를 들어서 Dockerfile 내부에
```
RUN apt-get update && \
    apt-get install -y \
        ca-certificates \
        git-core \
        vim-tiny zip unzip \
        libssl-dev \
        python3.9 python3-pip \ <- 이런식으로
        redis-server \
        mime-support && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/
```

**Reply by eunjungchoi** [reply]
저도 비슷한 문제를 겪었는데 알려주신 방법으로 해보니 설치가 되네요.   

(변경 후) 
python3.9 python3-pip \
python3.9-dev \ 

==>  python3 을 python3.9로 바꾸고,  python3.9-dev도 추가해주었습니다. 

 그리고 dockerfile 내에  
 RUN python3 -m pip install --no-cache-dir  -U -e . &&     rm -rf /root/.cache &&     rm -f /tmp/*.whl   
이 부분도  python3을 python3.9로 변경해주었습니다.

이렇게 변경했더니  설치가 잘 되네요. 감사드립니다.

**Reply by TeusEE** [reply]
저도 말씀해주신걸로 Dockerfile 설정 변경
python -> python3.9
python -m pip -> python3.9 -m pip
후에 docker-compose 로 image생성은 완료되었는데, 문제가 그 이후에 이제 test과정에서 Thread Start에서 에러가 나네요 ㅠㅠ
![image](https://user-images.githubusercontent.com/79349894/130353118-fbbe3776-d05c-43a6-a29f-1bb341536ae7.png)

![image](https://user-images.githubusercontent.com/79349894/130353128-46afe138-58bc-4671-86bd-7d0041d2c52f.png)

**Reply by youngjun0627** [reply]
Configuration에서 주소바꾸면 될거같아요
Dockerfile에 있는 webserver.conf 수정하면 될거같습니다

**Reply by TeusEE** [reply]
@youngjun0627 바쁘신데 도움주셔서 감사하니다.  하나만 더 여쭤볼게여 ㅠㅠ 혹시 말씀하신 주소가 뭘 말씀하신걸까요?
현재 backend.webui폴더 내 파일은 아래와 같구
![image](https://user-images.githubusercontent.com/79349894/130355532-de516372-3c13-4596-93ce-7f166e26ed83.png)

Dockerfile내 내용은 이렇습니다
```
FROM ubuntu:20.04
LABEL maintainer="jshin@lablup.com"

ENV DEBIAN_FRONTEND noninteractive
RUN apt-get update && \
    apt-get install -y \
        ca-certificates \
        git-core \
        vim-tiny zip unzip \
        python3.9 python3-pip \
        libssl-dev \
        redis-server \
        mime-support && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/

ENV PYTHONUNBUFFERED=1 \
    PATH=/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
    LANG=en_us.UTF-8

RUN python3.9 -V

RUN git clone https://github.com/lablup/backend.ai-webserver.git /webserver
WORKDIR "/webserver"
RUN python3.9 -m pip install --no-cache-dir  -U -e . && \
    rm -rf /root/.cache && \
    rm -f /tmp/*.whl
RUN update-alternatives --install /usr/bin/python python /usr/bin/python3.9 2
RUN redis-server --daemonize yes

COPY ./build/rollup /webserver/src/ai/backend/web/static

ENV BACKEND_ENDPOINT_TYPE=api
ENTRYPOINT redis-server --daemonize yes && python3.9 -m ai.backend.web.server -f /webserver.conf
```

혹시몰라서 webserver.conf도 첨부합니다!(backend.ai-webserver에 있는거 cp로 복사해서 넣어놨습니다)
[webserver.conf.txt](https://github.com/lablup/contribution-academy-2021/files/7027339/webserver.conf.txt)

**Reply by youngjun0627** [reply]
Webserver.conf를 바꾸셔야 할거에요. 주소 배정이 안된다고 뜨는거 보니깐 해당 주소포트를 가진 서버가 이미 실행중이거나 다른 프로세스가 실행중인거 같아요

**Reply by youngjun0627** [reply]
아마 제 생각에는 다른 컨테이너 실행중이신데 또 돌리신거같아요.
혹시 `docker ps -a`를 통해서 컨테이너 중복돼서 실행중인게 있다면 `docker rm`을 통해 없애고 다시 실행시켜보세요

**Reply by TeusEE** [reply]
말씀해주신걸로 확인해서 docker 컨테이너들의 포트번호 확인해봤습니다.
![image](https://user-images.githubusercontent.com/79349894/130360216-2a3788fc-9996-435c-a3fd-3eca6fafb576.png)


그래서, 사용중인 서버, Redis포트가 겹치지 않게 408X, 1808X으로 바꿨는데 동일하게 에러가 발생하네요 ㅠ
왜이런까요;;
![image](https://user-images.githubusercontent.com/79349894/130360206-ecd65760-b0eb-4709-b211-cb1dc6e25c88.png)

**Reply by youngjun0627** [reply]
~~흠... 자세히는 모르겠는데 저는 일단 docker_build에 있는 conf파일로 했거든요. 내일 한번 @TeusEE 님꺼 재현해볼게요~~

제가 해결한 방법은 다음과 같습니다.
webserver.conf에서 redis 관련 포트번호를 해당 redis 서버에 맞게 바꿔주시면 되요.

before :
![image](https://user-images.githubusercontent.com/57058726/130364776-fe45db29-dd3e-4e91-bb71-a472f07b851c.png)

after:
![image](https://user-images.githubusercontent.com/57058726/130364889-1473a8f5-fb5f-4c57-a93e-400b39de1c0b.png)

추가로 @TeusEE 님께서 보내주신 Dockerfile같은 경우 저는 에러가 나더라구요. 그래서 docker-compose.yml 파일이랑 Dockerfile을 수정해서 해결했는데, 이 부분에 대해서는 따로 문제가 없으신거 같으니 생략하겠습니다

**Comment by inureyes** [comment]
docker-compose 는 개발용으로 셋팅된거라 오래 업데이트를 하질 않았네요... 수정후에 PR 부탁드립니다 ㅎㅎ

**Reply by youngjun0627** [reply]
@inureyes 저번 PR과제처럼 우선 fork한 [개인레포 PR](https://github.com/youngjun0627/backend.ai-webui/pull/1)에 작성하였습니다. 피드백 받을수 있을까요?
